src/libos/crates/async-rt/src/sched/scheduler.rs

```diff
@@ -0,0 +1,283 @@
+    injector: Injector,
+    scheduling_epochs: AtomicU64,
+    rebalance_lock: Mutex<u64>,
+    rebalance_interval: u64,
+    parks: Arc<Parks>,
+}
+
+impl Scheduler {
+    const REBALANCE_BASE_INTERVAL: u64 = 32;
+
+    pub fn new(parallelism: u32, parks: Arc<Parks>) -> Self {
+        let workers = (0..parallelism).map(|_| Worker::new()).collect();
+        let injector = Injector::new();
+        let scheduling_epochs = AtomicU64::new(0);
+        let rebalance_lock = Mutex::new(0);
+        let rebalance_interval = Self::REBALANCE_BASE_INTERVAL * parallelism as u64;
+        Self {
+            parallelism,
+            workers,
+            injector,
+            scheduling_epochs,
+            rebalance_lock,
+            rebalance_interval,
+            parks,
+        }
+    }
+
+    /// Get a runnable task for specific thread.
+    pub fn get_runnable_task_for(&self, thread_id: usize) -> Option<Arc<Task>> {
+        // Increase scheduling_epochs in each scheduling.
+        let cnt = self.inc_scheduling_epochs();
+        // Try to do rebalance every once in a while.
+        if cnt % self.rebalance_interval == 0 {
+            self.try_rebalance_workload();
+        }
+
+        // Try to get a task from the worker.
+        match self.workers[thread_id].pop() {
+            Some(task) => {
+                // Update worker's latency.
+                // The current latency is calculated by the wait time of the task.
+                let latency = self.scheduling_epochs() - task.sched_info().scheduling_epochs();
+                self.workers[thread_id].update_latency(latency);
+                Some(task)
+            }
+            None => {
+                // Update worker's latency.
+                // We think the current latency is 0 when the worker is idle.
+                self.workers[thread_id].update_latency(0);
+                None
+            }
+        }
+    }
+
+    /// Schedule a new task.
+    pub fn schedule(&self, task: Arc<Task>) {
+        let thread_id = self.pick_thread_for(&task, None);
+        self.insert_task(task, thread_id);
+    }
+
+    /// Re-schedule a old task.
+    pub fn re_schedule(&self, task: Arc<Task>) {
+        let thread_id = if task.has_remained_budget() {
+            let last_thread_id = task.sched_info().last_thread_id() as usize;
+            let affinity = task.sched_info().affinity().read();
+            if affinity.get(last_thread_id) {
+                // Fast path: just use last_thread_id.
+                last_thread_id
+            } else {
+                // Slow path: the affinity is changed, need pick new thread.
+                drop(affinity);
+                self.pick_thread_for(&task, None)
+            }
+        } else if task.is_yield_loop() {
+            // Slow path: the budget has run out, pick new thread.
+            // We will forcely push the task to low priority queue since it's a yield loop.
+            self.pick_thread_for(&task, Some(SchedPriority::Low))
```

Comment:
No, only when the budget is zero, a yield-loop task will be scheduled as a low priority task.
When the task still has budget, it will be scheduled according to the priority when created